pilot/pkg/networking/grpcgen/grpc.go

```diff
@@ -0,0 +1,273 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package grpcgen
+
+import (
+	"net"
+	"strconv"
+	"strings"
+
+	xdsapi "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	envoycore "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	envoy_api_v2_route "github.com/envoyproxy/go-control-plane/envoy/api/v2/route"
+	v2 "github.com/envoyproxy/go-control-plane/envoy/config/filter/network/http_connection_manager/v2"
+	envoy_config_listener_v2 "github.com/envoyproxy/go-control-plane/envoy/config/listener/v2"
+	gogo "github.com/gogo/protobuf/proto"
```

Comment:
Please do not include gogo for xDS code. Something is guaranteed to go wrong.

Reply:
Agree - I'm still looking at how to fix the root problem, which is that envoyproxy/go-control-plane is not using gogo, while everything else in K8S and Istio is. And K8S is not going to change any time soon.

Until I find a way - and there are few options that seem viable - we need to support both in any generator that an return our high-level APIs ( like MCP ).

Reply:
It should at the very least be a separate file that imports gogo. k8s uses OpenAPIs AFAIK.

Reply:
I'll move the MCP part to separate file anyways - but it'll still need to use both, it needs to convert from the 2 'any' styles. 

K8S uses gogo/protobuf in the generated library - they do use OpenAPI as well, but k8s.io/apimachinery and the libraries we use gogo. 

So far it seems for serialization and deserialization it is safe to have protos generated with both gogo and golang - just like you can serialize with json and yaml. The only issue I see is in the gRPC service - and there is a solution there that will also 1/2 the number of goroutines we use, I'll explore it in a separate PR.

Reply:
I think the modules should produce standard Any. That way you can hide gogoproto in the module, and convert Any directly on output. Trying to change envoy or grpc or anything of google for that matter to use gogoproto is a path of extreme misery.

Reply:
I think the 'standard' for Istio and K8S is gogo. 

Control plane has used gogo - switched to golang - but we can switch back. The argument that grpc 'forces us' to use golang is not valid, there are ways around.

Reply:
I never heard of k8s exposing gogo. They use a forked version internally. Given that proto3 is undergoing a spec change, gogo is unlikely to keep up with the evolution of protobuf.